R/pmcmc.R

```diff
@@ -0,0 +1,609 @@
+
+    # propose new parameters
+    prop_pars <- propose_parameters(curr_pars,
+                                    proposal_kernel,
+                                    vars$range$discrete,
+                                    vars$range$min,
+                                    vars$range$max)
+
+    ## calculate proposed prior / lhood / posterior
+    prop_lprior <- calc_lprior(prop_pars, lprior_funcs)
+    prop_ll <- filter$run2(n_particles,
+                           save_history = FALSE,
+                           index = vars$index,
+                           pars = prop_pars,
+                           run_params = run_params)
+    prop_lpost <- prop_lprior + prop_ll
+
+    # calculate probability of acceptance
+    accept_prob <- exp(prop_lpost - curr_lpost)
+
+    # MH step
+    if (runif(1) < accept_prob) {
+      # update parameters and calculated likelihoods
+      curr_pars <- prop_pars
+      curr_lprior <- prop_lprior
+      curr_ll <- prop_ll
+      curr_lpost <- prop_lpost
+    }
+
+    # record results
+    res[iter, ] <- c(curr_pars,
+                     curr_lprior,
+                     curr_ll,
+                     curr_lpost)
+
+    if (output_proposals) {
+      proposals[iter, ] <- c(prop_pars,
+                             prop_lprior,
+                             prop_ll,
+                             prop_lpost,
+                             min(accept_prob, 1))
+    }
+
+  }
+
+  res <- as.data.frame(res)
+
+  coda_res <- coda::as.mcmc(res)
+  rejection_rate <- coda::rejectionRate(coda_res)
+  ess <- coda::effectiveSize(coda_res)
+
+  out <- list("results" = as.data.frame(res),
+              "acceptance_rate" = 1 - rejection_rate,
+              "ess" = ess)
+
+  if (output_proposals) {
+    proposals <- as.data.frame(proposals)
+    out$proposals <- proposals
+  }
+
+ class(out) <- "mcstate_pmcmc"
+ out
+}
+
+calc_lprior <- function(pars, pars_lprior) {
+  lprior <- 0
+  for (par in names(pars)) {
+      lprior <- lprior + pars_lprior[[par]](pars)
+  }
+  names(lprior) <- NULL
+  lprior
+}
+
+propose_parameters <- function(pars, proposal_kernel, pars_discrete,
+                               pars_min, pars_max) {
+
+  ## proposed jumps are normal with mean pars and sd as input for parameter
+  jumps <- pars + drop(rmvnorm(n = 1, sigma = proposal_kernel))
+
+  # discretise if necessary
+  jumps[pars_discrete] <- round(jumps[pars_discrete])
+  # reflect proposal if it exceeds upper or lower parameter boundary
+  jumps <- reflect_proposal(x = jumps,
+                          floor = pars_min,
+                          cap = pars_max)
+  jumps
+}
+
+## create function to reflect proposal boundaries at pars_min and pars_max
+# this ensures the proposal is symetrical and we can simplify the MH step
+reflect_proposal <- function(x, floor, cap) {
+  interval <- cap - floor
+  abs((x + interval - floor) %% (2 * interval) - interval) + floor
+}
+
+# Check input data.frame
+mcmc_validate_range <- function(range) {
```

Comment:
Is it worth trying to reduce the overlap here with the same function in grid? They look like they can share a common core

Reply:
Yes, especially as I started by copying that function